### ISSUE TITLE
ItemGroup item condition with metadata doesn't work in Project

### ISSUE THREAD

**Issue by jp2masa** [Microsoft/msbuild#3479]
`ItemGroup` with an item which has a condition with item metadata doesn't work in `Project`, but it works in a `Target`.

#### Example `ItemGroup`
```xml
<ItemGroup>
    <ContentX Include="@(Content)" Condition="'%(Content.X)' == 'abc'" />
</ItemGroup>
```

This works in a `Target`, but not in `Project`.

#### Command line
```
msbuild
```
or
```
dotnet build
```

### Expected  behavior
It uses the metadata to evaluate the condition.

### Actual behavior
It throws an exception:
```
error MSB4191: The reference to custom metadata "X" at position 1 is not allowed in this condition "'%(Content.X)' == 'abc'".
```

### Environment data
`msbuild /version` output:
15.7.179.6572

OS info:
Windows 10

If applicable, version of the tool that invokes MSBuild (Visual Studio, dotnet CLI, etc):
- Visual Studio: 15.7.4.
- .NET CLI: 2.1.300.

**Comment by TFTomSun** [comment]
@jp2masa 
There's another way to achieve it outside of targets:

```xml
      <ContentX Include="@(Content->HasMetadata('X')->WithMetadataValue('X','abc'))"/>
```
Documentation: https://docs.microsoft.com/de-de/visualstudio/msbuild/item-functions?view=vs-2019

It's important to check/filter first whether X is defined with HasMetadata, because WithMetadataValue throws an error, if the give meta value is not defined.

**Comment by am11** [comment]
Thanks for the alternative solution. I just realized that its negation would be:

```csproj
      <ContentX Include="@(Content)" Exclude="@(Content->HasMetadata('X')->WithMetadataValue('X','abc'))"/>
```
because:
```csproj
      <ContentX Include="@(!Content->HasMetadata('X')->WithMetadataValue('X','abc'))"/>
```
this will (silently) set the value to `@(!Content->HasMetadata('X')->WithMetadataValue('X','abc'))` without raising any errors (which is a bit counterintuitive given the other similar places where the exclamation mark is allowed).

**Comment by hknielsen** [comment]
I have same issue where I need to filter based on built-in metadata. Works when that same ItemGroup are inside a Task.
Ie.
```
error MSB4190: The reference to the built-in metadata "RootDir" at position 1 is not allowed in this condition "'%(RootDir)' == '$(MSBuildProjectExtensionsPath)'"
```

**Comment by binki** [comment]
Is this actually supposed to work? I don’t know what the actual limitations are supposed to be, but top-level `ItemGroup` seems to be intentionally feature-limited for performance. The actual expansion may happen after parsing whereas executing a condition based on metadata would require expansion to be performed during parsing unless things are made a lot more complicated…?

**Comment by danmoseley** [comment]
It's always been this way -- outside of Targets, each element is evaluated as a single entity, whereas syntax like ` Condition="'%(Content.X)' == 'abc'"` implies "batching" -- evaluating for each distinct bucket, which only happens inside of Targets. Later, filtering like `->WithMetadataValue('X','abc')` became possible.

I don't know of a fundamental reason why batching can't happen outside of targets, at least on Conditions, if someone was willing to do all the work and prove that performance wasn't negatively impacted.

**Comment by svew** [comment]
@danmoseley The official MSBuild documentation _frequently_ shows the bugged example as working code, I'm really hoping that at one point this was meant to work correctly. See this example on the [Use MSBuild items and item types](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-items?view=vs-2022#BKMK_Transforming) page:

```xml
<Target Name="ProcessCultureResources">
    <ItemGroup>
        <CultureResource Include="@(EmbeddedResource)"
            Condition="'%(EmbeddedResource.Culture)' != ''">
            <TargetDirectory>%(EmbeddedResource.Culture) </TargetDirectory>
        </CultureResource>
    </ItemGroup>
</Target>
```

**Comment by binki** [comment]
@svew
> The official MSBuild documentation _frequently_ shows the bugged example as working code, so I'm really hoping that you're wrong about it having always worked this way, otherwsie it's a bad omen for the accuracy of MSBuild docs generally... See this example on the [Use MSBuild items and item types](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-items?view=vs-2022#BKMK_Transforming) page:

The example you mention includes `<Target/>`. This issue is also sort of a feature request; docs bugs should probably be filed as separate issues (but, again, I think that example is fine).

**Comment by svew** [comment]
> @svew
> > The official MSBuild documentation _frequently_ shows the bugged example as working code, so I'm really hoping that you're wrong about it having always worked this way, otherwsie it's a bad omen for the accuracy of MSBuild docs generally... See this example on the [Use MSBuild items and item types](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-items?view=vs-2022#BKMK_Transforming) page:
> 
> The example you mention includes `<Target/>`. This issue is also sort of a feature request; docs bugs should probably be filed as separate issues (but, again, I think that example is fine).

You are correct, it doesn't actually refer to this item, I deleted the comment once I realized

**Comment by maridematte** [comment]
Triage: We probably will not change this in the near future. But we can update the documentation for the error so it is easier to find a solution when you run into this.
